File: backend/src/SwiftShopper.Domain/Entities/RequestItem.cs
namespace SwiftShopper.Domain.Entities;

public class RequestItem
{
    public required string Name { get; init; }

    public string Unit { get; init; } = string.Empty;

    public string Description { get; init; } = string.Empty;

    public decimal Price { get; init; }

    public int Quantity { get; init; }

    public decimal? MaxPrice { get; init; }
}

File: lib/features/create_request/models/request_item.dart
class RequestItem {
  const RequestItem({
    required this.id,
    this.name = '',
    this.unit = '',
    this.description = '',
    this.price = 0.0,
    this.quantity = 1,
    this.maxPrice,
  });

  final String id;
  final String name;
  final String unit;
  final String description;
  final double price;
  final int quantity;
  final double? maxPrice;

  double get subtotal => price * quantity;

  RequestItem copyWith({
    String? id,
    String? name,
    String? unit,
    String? description,
    double? price,
    int? quantity,
    double? maxPrice,
    bool clearMaxPrice = false,
  }) {
    return RequestItem(
      id: id ?? this.id,
      name: name ?? this.name,
      unit: unit ?? this.unit,
      description: description ?? this.description,
      price: price ?? this.price,
      quantity: quantity ?? this.quantity,
      maxPrice: clearMaxPrice ? null : (maxPrice ?? this.maxPrice),
    );
  }
}
